### PR TITLE
Implement bigtable::Table::AsyncCheckAndMutateRow

### LINKED ISSUE #1489: Implement bigtable::Table::AsyncCheckAndMutateRow
This function will implement an asynchronous version of `Table::ReadRow`. The function will have the following signature:

```C++
class google::cloud::bigtable::Table { public:
template <typename ReadRowFunctor, ...>
future<bool> AsyncCheckAndMutateRow(
    std::string row_key, Filter filter,
    std::vector<Mutation> true_mutations,
    std::vector<Mutation> false_mutations,
    CompletionQueue& cq);
};
```

As part of this work we need to:

- [ ] Implement the function.
- [ ] Implement an integration test for this function.
- [ ] Implement an example for this function.
- [ ] Execute the example as part of the CI build.

The main difference between the example and the integration test is that the latter uses pre-defined values and checks the resulting state against expected outputs, while the former takes (most) of its parameters from the command-line and simply prints out any state.

### PR DESCRIPTION
**Pull request by rsimion**
Fixes #1489 

Part of #2097

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/2382)

### CODE REVIEW

File: google/cloud/bigtable/examples/data_async_snippets.cc [hunk 1]
```diff
@@ -22,6 +22,7 @@
 
 namespace {
 namespace cbt = google::cloud::bigtable;
+const std::string MAGIC_ROW_KEY = "key-000005";
```
Comment (remyabel) [inline google/cloud/bigtable/examples/data_async_snippets.cc]: What's the agreed upon convention for constant naming here? Should this be `kMagicRowKey`?

Comment (rsimion) [reply google/cloud/bigtable/examples/data_async_snippets.cc]: I used the same format as in data_snippets.cc
@coryan would know better.

File: google/cloud/bigtable/examples/data_async_snippets.cc [hunk 2]
```diff
@@ -124,6 +125,47 @@ void AsyncBulkApply(cbt::Table table, cbt::CompletionQueue cq,
...
+          if (!row) {
+            throw std::runtime_error(row.status().message());
+          }
+
+          return google::cloud::Status();
+        });
+    final.get();
+  }
+  //! [async check and mutate]
+  (std::move(table), std::move(cq), argv[1]);
```
Comment (remyabel) [inline google/cloud/bigtable/examples/data_async_snippets.cc]: What do you think about using the [`ConsumeArg`](https://github.com/googleapis/google-cloud-cpp/blob/e8e461a3416d339cde9948fb2f785ce7dc22e885/google/cloud/bigtable/examples/data_snippets.cc#L39) function present in the other sample files for consistency?

Comment (rsimion) [reply google/cloud/bigtable/examples/data_async_snippets.cc]: I agree, but I do not want to start changing things around unless approved (@coryan).

File: google/cloud/bigtable/examples/data_async_snippets.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/bigtable/examples/data_async_snippets.cc]: We have been terribly inconsistent with naming constants, we need to make a pass cleaning them up.  Unfortunately the clang-tidy checks break if you change some of these, so I have been leaving them alone until we figure out the correct incantations for clang-tidy.

File: google/cloud/bigtable/examples/data_async_snippets.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/bigtable/examples/data_async_snippets.cc]: Another case where we should probably make a pass and remove all the `ConsumeArg` things.

File: google/cloud/bigtable/examples/data_async_snippets.cc (hunk 1 above)
Comment (remyabel) [reply google/cloud/bigtable/examples/data_async_snippets.cc]: I did attempt it a while ago. Something like the following should work in theory:

```
- { key: readability-identifier-naming.GlobalConstantPrefix,    value: k }
- { key: readability-identifier-naming.GlobalConstantCase,     value: CamelCase }
```

If I recall, after I applied this pattern to the existing check options, it started complaining about about the include guards.

**Pull request merged**